daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1013,33 +1023,25 @@ object SBuiltin {
       }
       val coinst =
         machine.ptx
-          .lookupCachedContract(coid)
+          .lookupLocalContract(coid)
           .getOrElse(
             coid match {
-              case V.AbsoluteContractId.V1(discriminator, _)
-                  if machine.ptx.localContracts.isDefinedAt(
-                    V.AbsoluteContractId.V1(discriminator)) =>
-                crash(
-                  s"The local contract discriminator $discriminator is not fresh in the transaction")
               case acoid: V.AbsoluteContractId =>
                 throw SpeedyHungry(
                   SResultNeedContract(
                     acoid,
                     templateId,
                     machine.committers,
                     cbMissing = _ => machine.tryHandleException(),
-                    cbPresent = {
-                      coinst =>
-                        // Note that we cannot throw in this continuation -- instead
-                        // set the control appropriately which will crash the machine
-                        // correctly later.
-                        if (coinst.template != templateId) {
-                          machine.ctrl =
-                            CtrlWronglyTypeContractId(acoid, templateId, coinst.template)
-                        } else {
-                          machine.ptx = machine.ptx.cachedContract(coid, coinst)
-                          machine.ctrl = translateValue(machine, coinst.arg.value)
-                        }
+                    cbPresent = { coinst =>
+                      // Note that we cannot throw in this continuation -- instead
+                      // set the control appropriately which will crash the machine
+                      // correctly later.
+                      if (coinst.template != templateId) {
+                        machine.ctrl = CtrlWronglyTypeContractId(acoid, templateId, coinst.template)
+                      } else {
+                        importValue(machine, coinst.arg.value)
```

Comment:
`importValue` will throw if there's a clash of discriminators, but the comment a few lines above says that we must not throw here. This is not consistent.

Reply:
fix.